receiver/kafkareceiver/avro_unmarshaler.go

```diff
@@ -0,0 +1,184 @@
+	mapping      map[string]string
+}
+
+func newAVROLogsUnmarshaler() *avroLogsUnmarshaler {
+	return &avroLogsUnmarshaler{}
+}
+
+func (a *avroLogsUnmarshaler) Encoding() string {
+	return avroEncoding
+}
+
+func (a *avroLogsUnmarshaler) Unmarshal(data []byte) (plog.Logs, error) {
+	avroLog, err := a.deserializer.Deserialize(data)
+	if err != nil {
+		return plog.NewLogs(), fmt.Errorf("failed to deserialize avro log: %w", err)
+	}
+
+	return a.mapLog(avroLog), nil
+}
+
+func (a *avroLogsUnmarshaler) mapLog(avroLog map[string]interface{}) plog.Logs {
+	logs := plog.NewLogs()
+	resourceLogs := logs.ResourceLogs().AppendEmpty()
+	resourceAttributes := resourceLogs.Resource().Attributes()
+	scopeLogs := resourceLogs.ScopeLogs().AppendEmpty()
+	logRecord := scopeLogs.LogRecords().AppendEmpty()
+	logRecordAttributes := logRecord.Attributes()
+
+	for sourceField, target := range a.mapping {
+		switch {
+		case isResourceAttribute(target):
+			targetField := strings.TrimPrefix(target, resourceAttributesPrefix)
+			mapAttribute(avroLog, sourceField, &resourceAttributes, targetField)
+		case isLogRecordAttribute(target):
+			targetField := strings.TrimPrefix(target, logRecordAttributesPrefix)
+			mapAttribute(avroLog, sourceField, &logRecordAttributes, targetField)
+		case target == bodyField:
+			logRecord.Body().SetStr(avroLog[sourceField].(string))
+		case target == timestampField:
+			timestamp := avroLog[timestampField].(time.Time)
+			logRecord.SetTimestamp(pcommon.NewTimestampFromTime(timestamp))
+		case target == severityTextField:
+			logRecord.SetSeverityText(avroLog[sourceField].(string))
+		case target == severityNumberField:
+			severityNumber, ok := avroLog[sourceField].(int32)
+			if !ok {
+				severityNumber = 0
+			}
+			logRecord.SetSeverityNumber(plog.SeverityNumber(severityNumber))
+		}
+	}
+
+	return logs
+}
+
+func isResourceAttribute(targetField string) bool {
+	return strings.HasPrefix(targetField, resourceAttributesPrefix)
+}
+
+func isLogRecordAttribute(targetField string) bool {
+	return strings.HasPrefix(targetField, logRecordAttributesPrefix)
+}
+
+func mapAttribute(avroLog map[string]interface{}, sourceField string, attributes *pcommon.Map, targetField string) {
+	sourceValue := avroLog[sourceField]
+	if sourceValue != nil {
+		setAttribute(attributes, targetField, sourceValue)
+	}
+}
+
+func setAttribute(attributes *pcommon.Map, key string, value interface{}) {
+	attribute := attributes.PutEmpty(key)
+	_ = attribute.FromRaw(value)
+}
+
+func (a *avroLogsUnmarshaler) Init(schemaURL string, mapping map[string]string) error {
+	a.mapping = mapping
+
+	var schemaDeserializer avroDeserializer
+	var err error
+
+	if strings.HasPrefix(schemaURL, avroSchemaPrefixFile) {
+		schemaFilePath := strings.TrimPrefix(schemaURL, avroSchemaPrefixFile)
+		schemaFilePath = filepath.Clean(schemaFilePath)
+		schemaDeserializer, err = newAVROFileSchemaDeserializer(schemaFilePath)
+	} else {
+		// TODO: schema registry deserializer
```

Comment:
Besides reading a schema from file, it can also be read from a registry like Confluents Schema Registry. Therefore a magic byte is added in front of the message representing the schema id that needs to be used to receive the schema. I intended to also provide an implementation to support this in the future.

Reply:
I suggest leaving the schema registry for a new issue since that will have broader impact on the project since we can potentially pull it out and use it in more places.